src/balance.rs

```diff
@@ -26,41 +42,186 @@ impl<T: Config> From<Error<T>> for StatusCode {
+        &self,
+        account: &AccountIdOf<T>,
+        amount: &BalanceOf<T>,
+    ) -> DispatchResult {
+        if *amount < T::Currency::free_balance(account) {
+            Ok(())
+        } else {
+            Err(Error::<T>::InsufficientBalance.into())
+        }
+    }
+
+    /// Does a state checking on initial state of cheques with current state.
+    fn cmp_with_initial_state(&self) -> DispatchResult {
+        use sp_std::cell::Ref;
+
+        let cheques: Ref<HashMap<AccountIdOf<T>, BalanceOf<T>>> = self.cheques.borrow();
+
+        let mut sum_initial = BalanceOf::<T>::zero();
+        let mut sum_cheques = BalanceOf::<T>::zero();
+        self.initial_state.values().for_each(|b| sum_initial += *b);
+        cheques.values().for_each(|b| sum_cheques += *b);
+
+        if sum_initial == sum_cheques {
+            Ok(())
+        } else {
+            Err(DispatchError::Corruption)
         }
     }
 }
 
-impl<T: Config> Default for BalanceAdapter<T> {
+impl<T: Config> Default for BalanceAdapter<T>
+where
+    AccountIdOf<T>: core::cmp::Eq + core::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     fn default() -> Self {
         Self::new()
     }
 }
 
-impl<T: Config> BalanceHandler for BalanceAdapter<T> {
+impl<T: Config> BalanceHandler for BalanceAdapter<T>
+where
+    AccountIdOf<T>: core::cmp::Eq + core::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     type Error = StatusCode;
 
     fn transfer(
         &self,
-        _src: AccountAddress,
-        _dst: AccountAddress,
-        _cheque_amount: u128,
+        src: AccountAddress,
+        dst: AccountAddress,
+        cheque_amount: u128,
     ) -> Result<bool, Self::Error> {
-        // TODO(rqnsom): This won't break the MoveVM unless the balance module is accessed.
-        unimplemented!()
+        let from = Pallet::<T>::to_native_account(&src)?;
+        let to = Pallet::<T>::to_native_account(&dst)?;
+        let amount = BalanceOf::<T>::from(cheque_amount);
+
+        let mut cheques = self.cheques.borrow_mut();
+
+        let src_balance = cheques.entry(from).or_insert(BalanceOf::<T>::zero());
+        if *src_balance < amount {
+            return Err(StatusCode::INSUFFICIENT_BALANCE);
+        }
+        *src_balance -= amount;
+
+        if let Some(dst_balance) = cheques.get_mut(&to) {
+            *dst_balance += amount;
+        } else {
+            cheques.insert(to, amount);
+        }
+
+        Ok(true)
     }
 
-    fn cheque_amount(&self, _account: AccountAddress) -> Result<u128, Self::Error> {
-        // TODO(rqnsom): This won't break the MoveVM unless the balance module is accessed.
-        unimplemented!()
+    fn cheque_amount(&self, account: AccountAddress) -> Result<u128, Self::Error> {
+        let acc = Pallet::<T>::to_native_account(&account)?;
+        let cheques = self.cheques.borrow();
+        let balance = cheques
+            .get(&acc)
+            .ok_or(StatusCode::SENDING_ACCOUNT_DOES_NOT_EXIST)?;
```

Comment:
Returning zero should still be the correct approach here (check my old comment about this one - I think I resolved it accidentally... sorry :D)

Reply:
Done :blush: